운동 시작 전 타이머 화면 구현

jeonyeohun opened the pull request:
### 📕 Issue Number

Close #10

### 📙 작업 내역

> 구현 내용 및 작업 했던 내역

- [x] 타이머 구현
- [x] 타이머 변경 애니메이션 구현
- [x] 테스트 작성
- 구현결과

https://user-images.githubusercontent.com/46087477/140120226-9e488794-985e-414d-82f3-76b60275c1ad.MP4

- 테스트 커버리지
<img width="811" alt="스크린샷 2021-11-04 오전 1 49 07" src="https://user-images.githubusercontent.com/46087477/140120638-060c64b0-9a14-47ca-bbec-2f9a4d2d73d5.png">

### 📘 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트


### 📋 체크리스트

- [x] Merge 하는 브랜치가 올바른가?
- [x] 코딩컨벤션을 준수하는가?
- [x] PR과 관련없는 변경사항이 없는가?
- [x] 내 코드에 대한 자기 검토가 되었는가?
- [x] 변경사항이 효과적이거나 동작이 작동한다는 것을 보증하는 테스트를 추가하였는가?
- [x] 새로운 테스트와 기존의 테스트가 변경사항에 대해 만족하는가?
      <br/>

### 📝 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점
- 전체적으로 다음과 같은 흐름으로 설계하고 구현했습니다.
```
- View 에서는 ViewDIdLoad에 대한 이벤트를 ViewModel에 전달합니다. 
- ViewModel은 이벤트를 받은 뒤 UseCase의 시작 함수를 호출합니다.
- UseCase는 타이머를 시작하고 진행된 경과시간을 업데이트합니다.
- ViewModel은 경과시간을 옵저빙하고 업데이트 될 때마다 스트림으로 값을 받아 준비운동시간(3초)에서 값을 빼서 남은 시간을 텍스트로로 포맷합니다.
- View는 ViewModel의 남은시간을 옵저빙하고 남은시간이 업데이트되면 애니메이션과 함께 뷰를 업데이트합니다.
- ViewModel은 추가적으로 경과시간을 옵저빙하면서 남은 시간이 0이하인지에 따라 뷰로 화면 전환 플래그를 방출합니다.
- View는 true값을 스트림을 통해 받으면 화면을 전환합니다.
```
- 0초
저희가 기획한 사항은 3, 2, 1만 보이는 건데 화면 전환 플래그를 확인하기 위해서 0초를 넣어두었습니다. 다음화면이랑 이전화면 완성되면 연결하면서 수정하겠습니다!

- 타이머 정확도? 문제
타이머를 1초 간격으로 블록을 실행하도록 처리했는데, 실기기에서는 문제가 없는 것 같지만 시뮬레이터에서는 정확히 1초가 아니라 1초 내로 조금 일관성 없게 블록이 실행됩니다. 시뮬레이터 문제인지 코드의 문제인지 정확하지가 않아서 실기기와 시뮬에서 모두 테스트해보시고 알려주시면 감사할 것 같습니다!

- 라이브러리 추가
RxTest를 테스트 타겟에 대한 라이브러리로 추가했습니다. 일단 제가 만든 타겟에만 적용되도록 해놔서 다른 파일도 테스트하려면 Pod파일에 추가를 해줘야할 것 같습니다. 전역으로 모든 테스트 타겟에 적용할 수 있는 방법을 찾아보겠습니다!

<br/><br/>

jeonyeohun committed:
[Feat] 운동 시작 전 타이머 뷰 구현

- 뷰(이벤트) -> 뷰모델(남은시간으로 포맷변경) -> 유즈케이스(타이머) 의 구조로 설계했습니다.
- 애니메이션은 폰트를 작은 사이즈로 만들었다가 설정된 크기로 transform 하도록 구현했습니다.
- 아직 다음 화면 전환에 대한 추가적인 코든는 TODO로 마킹한 코드영역을 수정하면 될 것 같습니다.

jeonyeohun committed:
[Fix] 파일이름 오타 수정

jeonyeohun committed:
[Refactor] 테스트 가능한 구조를 위해 프로토콜로 의존성 분리

jeonyeohun committed:
[Fix] 내비게이션 관련 output 변경

- 기존에 Void 타입이였던 output을 bool 타입으로 변경해서 내비게이션 여부를 확인할 수 있도록 수정했습니다.

jeonyeohun committed:
[Chore] RxTest 라이브러리 추가

- Rx 코드 테스트를 위한 라이브러리를 추가했습니다.

jeonyeohun committed:
[Refactor] 프로토콜, 구현체 파일 분리

jeonyeohun committed:
[Feat] 테스트를 위한 UseCase Mock객체 정의 및 테스트 추가

- Mock 객체에서는 타이머를 사용하지 않고, 1, 2, 3을 순차적으로 스트림에 방출하도록 정의했습니다.
- 이 값에 따라 뷰 모델에서 남은 시간을 계산하고 텍스트로 변환해서 뷰에 표시하거나 화면 전환 여부를 뷰에 방출하도록 했습니다.
- 테스트를 통해 뷰에 도착하는 값들이 올바른지 확인했습니다.

jeonyeohun committed:
[Refactor] 코딩컨벤션 준수여부 검토 및 수정

- self, 줄바꿈 등을 검토하고 수정했습니다.

jeonyeohun committed:
[Refactor] import 순서변경

lee-yujinn commented on MateRunner/MateRunner/Presentation/HomeScene/ViewModel/RunningPreparationViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+		@BehaviorRelayProperty var navigateToNext: Bool?
+	}
+
+	let useCase: RunningPreparationUseCase
```
여러개의 usecase를 가지고 있을 경우를 생각해서 이름을 좀 더 명확히 하면 좋을 것 같습니다!!

lee-yujinn commented on MateRunner/MateRunner/Domain/UseCase/Home/DefaultRunningPreparationUseCase.swift:
```diff
@@ -0,0 +1,24 @@
+	var timeSpent = BehaviorSubject(value: 0)
+	private let maxTime = 3
+	
+	func execute() {
```
excute() 라는 메서드명이 더 명확하면 좋을것 같아요~

jeonyeohun replied:
executeTimer 말씀이시죠? 수정하겠슴다~

lee-yujinn commented on MateRunner/Tests/RunningPreparationViewModelTests/RunningPreparationViewModelTests.swift:
```diff
@@ -0,0 +1,77 @@
+			.next(10, false),
+			.next(10, false),
+			.next(10, true)
+		])
```
테스트코드까지....👍

lee-yujinn commented on MateRunner/MateRunner/Presentation/HomeScene/ViewController/RunningPreparationViewController.swift:
```diff
@@ -0,0 +1,67 @@
+import RxSwift
+import SnapKit
+
+class RunningPreparationViewController: UIViewController {
```
final!

jeonyeohun committed:
[Refactor] 코드리뷰 반영

- 몇가지 코딩 스타일을 수정했습니다.(final 붙이기)
- 타임오버체킹을 뷰 모델이 아닌 유즈케이스에서 수행하도록 했습니다.
- 유즈케이스 변경에 따라 프로토콜과 테스트케이스를 수정했습니다.

jeonyeohun committed:
[Refactor] 유즈케이스 이름 수정

- 이후에 다른 유즈케이스가 추가될 수 있다는 리뷰를 반영하여 구체적인 이름으로 수정하였습니다.

jeonyeohun committed:
[Chore] 유즈케이스 프로토콜 디렉토리 추가 후 프로토콜 파일 이동
